Isolated synchronous deinit

nickolas-pohilets opened the pull request:
Implementation of the [Isolated asynchronous deinit](https://github.com/nickolas-pohilets/swift-evolution/blob/mpokhylets/isolated-deinit/proposals/NNNN-isolated-synchronous-deinit.md)

~Still pending:~
* ~I suspect that logic for adding attributes for inferred isolation is not correct, would like some input from the Core team about this.~
* ~Still need to integrate with distributed actors, waiting for https://github.com/apple/swift/pull/60050 to be merged.~
* ~Didn't copy changes to `BackDeployConcurrency` yet. I expect that I'll have to make changes based on review comments, and want to apply all changes to `BackDeployConcurrency` in one go. I would appreciate some advice for maintaining the two in sync.~
* ~Investigate fast path for self-isolated default actor's deinit~

nickolas-pohilets committed:
Runtime support for isolated deinit

DougGregor commented on lib/SILGen/SILGenDestructor.cpp:
```diff
@@ -213,6 +213,99 @@ void SILGenFunction::emitDeallocatingDestructor(DestructorDecl *dd) {
+  // If isolation was intoduced in ObjC code, then we assume that ObjC code also
+  // overrides retain/release to make sure that dealloc is called on the correct
+  // executor in the first place.
+  return firstIsolated->getClangNode().isNull();
```
I think there's an additional optimization we should consider here, although I'm not sure how tricky it is. When the deinitializer is for a "default actor", i.e., one that has not overridden `unownedExecutor`, we shouldn't need to hop to the executor because we know it's available. I wonder if we could contrive to immediately run the deallocating initializer rather than going through "perform on executor." The only thing I'm not sure of is how to ensure that `swift_task_isCurrentExecutor` will return true; perhaps we can temporarily set the thread-specific storage to the actor's executor, run the deallocating initializer code, then set it back.

nickolas-pohilets replied:
Implemented, please take a look

DougGregor reviewed (commented):
This is fantastic! A couple of comments and questions, both in design and implementation, but it's very much going in the right direction.

DougGregor commented:
> * Didn't copy changes to `BackDeployConcurrency` yet. I expect that I'll have to make changes based on review comments, and want to apply all changes to `BackDeployConcurrency` in one go. I would appreciate some advice for maintaining the two in sync.

`BackDeployConcurrency` is actually supposed to be frozen in time, and not updated with new features. For back-deployment of this feature, we'll have to bring up another statically-linked back-deployment library that would be linked into all problems that build with a deployment target that predates this work. @etcwilde is starting to bring up such a library to back-deploy other concurrency-related fixes. Once that's in place, this PR can build on that work.

nickolas-pohilets commented on stdlib/public/Concurrency/MainActor.swift:
```diff
@@ -70,6 +70,8 @@ import Swift
     _enqueueOnMain(job)
   }
+  
+  nonisolated deinit {}
```
Not needed. Leftover from early experiments. Will remove.

nickolas-pohilets commented on test/Concurrency/deinit_isolation.swift:
```diff
@@ -0,0 +1,405 @@
+// RUN: %target-swift-frontend -parse-as-library -emit-silgen -verify %s
+// RUN: %target-swift-frontend -parse-as-library -emit-silgen -DSILGEN %s | %FileCheck %s
+// RUN: %target-swift-frontend -parse-as-library -emit-silgen -DSILGEN %s | %FileCheck -check-prefix=CHECK-SYMB %s
```
It does not really check the generated SIL, it only checks that symbol exists, as an evidence of isolation.

nickolas-pohilets commented on stdlib/public/Concurrency/Task.cpp:
```diff
@@ -1394,6 +1394,65 @@ swift_task_createNullaryContinuationJobImpl(
+  AdHocJob(JobPriority priority, void *context, AdHocWorkFunction *work)
+      : Job({JobKind::AdHoc, priority}, &process), Context(context),
+        Work(work) {
+    TaskLocal::copyTo(&Local, nullptr);
```
For GAIT, I expect that vast majority of releases to happen within isolating actor, so I'm not very concerned about performance costs of the slow path. But for actors, that's probably not true. Maybe what you mentioned [below](https://github.com/apple/swift/pull/60057#discussion_r922585619) can help to reduce switching for actors.

ktoso replied:
I would not expect to have task locals available in a deinit, and especially since we don't guarantee WHICH exact locals those will be, so basing any tracing infra on "object got deallocated and must have right locals" is going to be wrong at some point so not usable.

I'd say losing the locals is expected. For anything that needs things at "end of scope" (like tracing) the usual pattern is to wrap in a closure, not really rely on `deinit`.

nickolas-pohilets replied:
I think some task-locals can still be useful in the `deinit` even if you are not sure which task/thread performs the last release. For example, if you have a logger applied to entire app or a large subsystem, you can expect that it will be available in any task/thread launched within that scope. So it does not matter which one makes the last release, as long you are confident that last release happens within this coarse-grained scope, you can still meaningfully use logger's task local in the `deinit`.

nickolas-pohilets commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -4505,6 +4505,22 @@ namespace {
       result->setHasMissingVTableEntries(false);
       result->setMemberLoader(&Impl, 0);
 
+      // GetDestructorRequest does not trigger lazy member loading
+      // And typechecking may ask for destructor before member loading is
+      // triggered. Create deinit explicitly
```
Importing `dealloc` as two entities is intentional. When `dealloc` is imported as a function, it carries deprecation attribute. So at least the `dealloc` of the `NSObject` needs to be imported as a function.

I'll check about `GetDestructorRequest` and will get back to you.

nickolas-pohilets replied:
Currently `GetDestructorRequest` lives in `swiftAST`, while `Impl.createDeclWithClangNode` lives in `swiftClangImporter` and `swiftClangImporter` depends on `swiftAST`. So using `Impl.createDeclWithClangNode` in `GetDestructorRequest::evaluate()` would create a cyclic dependency between libraries.

nickolas-pohilets commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -3651,15 +3664,84 @@ static OverrideIsolationResult validOverrideIsolation(
     auto actor = value->getDeclContext()->getSelfNominalTypeDecl();
     assert(actor && "could not find the actor that 'self' is isolated to");
+
+    // Bootstrapping hack: force _Concurrency.MainActor.deinit() to be
+    // non-isolated even when importing from SDK's swiftmodule without
+    // `nonisolated` attribute.
```
It's not needed with updated runtime, but needed when building with hostlibs. From `Xcode-13.3.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/lib/swift/_Concurrency.swiftmodule/arm64e-apple-macos.swiftinterface`:

```swift
#if compiler(>=5.3) && $Actors && $GlobalActors
@_hasMissingDesignatedInitializers @available(macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0, *)
@globalActor final public actor MainActor : _Concurrency.GlobalActor {
  ...
  @objc deinit // explicit and not isolated, gets typechecked as self-isolated
}
#endif
```

nickolas-pohilets commented:
> `BackDeployConcurrency` is actually supposed to be frozen in time, and not updated with new features. For back-deployment of this feature, we'll have to bring up another statically-linked back-deployment library that would be linked into all problems that build with a deployment target that predates this work. @etcwilde is starting to bring up such a library to back-deploy other concurrency-related fixes. Once that's in place, this PR can build on that work.

I would really appreciate a document about how to work with `BackDeployConcurrency`. Could not find anything in the repo.

ktoso commented:
AFAICS the distributed adjustment looks good 👍 
We only `resignID` as before in the local branch, and the remote does only destroy fields properly, looks good to me.

nickolas-pohilets commented on test/Concurrency/Runtime/actor_deinit_escaping_self.swift:
```diff
@@ -0,0 +1,69 @@
+    }
+    let r = g.wait(timeout: .now() + .milliseconds(500))
+    expectEqual(r, .timedOut)
+    expectCrashLater(withMessage: "Assertion failed: (!oldState.getFirstJob() && \"actor has queued jobs at destruction\"), function destroy")
```
I'm not sure if that is even valid to test for assertions in stdlib as an expected behaviour. If stdlib is build without assertions, test will fail

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test

xwu commented:
@swift-ci please test

xwu commented:
@swift-ci please test

stephencelis commented:
I asked on the review thread, but maybe it'll get more visibility here. Any chance we can get a toolchain to test?

theblixguy commented:
@swift-ci please build toolchain

Jumhyn commented:
@swift-ci please build toolchain

Jumhyn commented:
@swift-ci please build toolchain macOS

Jumhyn commented:
@swift-ci please build toolchain

DougGregor commented:
@swift-ci please build toolchain macOS

DougGregor commented:
I think that swiftThreading error should be gone now. I saw it late last week in an unrelated pull request, and it's gone now. However, the conflict will prevent us from getting clean builds on this PR.

ktoso commented on include/swift/ABI/Task.h:
```diff
@@ -333,14 +333,15 @@ class AsyncTask : public Job {
   // ==== Task Local Values ----------------------------------------------------
 
-  void localValuePush(const HeapObject *key,
+  void localPushValue(const HeapObject *key,
```
Please don't rename method names here, this is `ABI/Task.h` after all

nickolas-pohilets replied:
Done

ktoso commented on include/swift/ABI/Task.h:
```diff
@@ -333,14 +333,15 @@ class AsyncTask : public Job {
                       const Metadata *valueType);
 
-  OpaqueValue *localValueGet(const HeapObject *key);
+  void localPushStop();
```
I'd like to use the term `Barrier` for those; so following the existing naming it would be `localValuePushBarrier`

nickolas-pohilets replied:
Done

ktoso commented on include/swift/ABI/TaskLocal.h:
```diff
@@ -32,62 +32,71 @@ class TaskGroup;
+    /// Stop-item that blocks further lookup.
+    /// Inserting stop-node allows to temporary disable all inserted task-local values in O(1),
+    /// while maintaining immutable linked list nature of the task-local values implementation.
+    Stop = 1,
```
I'd like to use the `Barrier` terminology for this operation. Or `LookupBarrier` if we wanted to be longer.

nickolas-pohilets replied:
Done

tshortli commented on test/ModuleInterface/SmokeTest.swiftinterface:
```diff
@@ -33,7 +32,7 @@ public class TestClass {
   public static var propWithNoAccessors: Int
 
-  // NEGATIVE-NOT: deinit
+  // CHECK: deinit
```
Why did the printing behavior change here?

tshortli replied:
Nevermind; I see elsewhere that the printing behavior was changed globally for deinit to accommodate the new modifier.

tshortli commented on lib/AST/Decl.cpp:
```diff
@@ -8883,10 +8901,10 @@ bool ConstructorDecl::isObjCZeroParameterWithLongSelector() const {
 }
 
-DestructorDecl::DestructorDecl(SourceLoc DestructorLoc, DeclContext *Parent)
+DestructorDecl::DestructorDecl(SourceLoc DestructorLoc, bool Async, SourceLoc AsyncLoc, DeclContext *Parent)
```
Nit: Run `git-clang-format` over the commits in the PR

nickolas-pohilets replied:
Done

tshortli commented on utils/gyb_syntax_support/AttributeKinds.py:
```diff
@@ -861,7 +867,7 @@ def __init__(self, name, swift_name=None):
 
     ContextualSimpleDeclAttribute('isolated', 'Isolated',
-                                  DeclModifier, OnParam,
+                                  DeclModifier, OnParam, OnDestructor,
```
Given that the parser needed to be updated to accept `isolated` on destructors, we need to be careful about interface printing. ASTPrinter.cpp needs to be updated with a new "feature" for isolated deinits and it needs to print the declaration of any `deinit` with `isolated` with a feature guard so that older development versions of the compiler will continue to be able to parse `swiftinterfaces` for modules that have adopted this feature. There should also be a test under the `tests/ModuleInterface/`subdirectory that exercises printing of isolated deinits and verifies that they are printed with the feature guards.

nickolas-pohilets replied:
`isolated` is an attribute, and there is already code in place in ASTPrinter which prints attributes, so it should work, but worth to add couple more tests cases. But I could not find which test tests ASTPrinter. Any examples to look at?

tshortli replied:
`tests/ModuleInterface/` contains examples of tests that exercise printing.

To be clearer about what I'm saying the problem will be: already existing compilers that do not have the changes from this PR will presumably reject this declaration when parsing a `.swiftinterface`:
```
class SomeClass {
  isolated deinit // error: 'isolated' may only be used on 'parameter' declarations
}
```
Therefore, to maintain compatibility with older compilers, we need to print the deinit declaration that has adopted this new capability something like this:
```
class SomeClass {
  #if $IsolatedDeinit
  isolated deinit
  #else
  deinit
  #endif
}
```
This can be achieved with a suppressible language feature (see https://github.com/apple/swift/blob/main/include/swift/Basic/Features.def)

tshortli replied:
Perhaps, yes. But that gets tricky because other references to the class elsewhere in the interface would also need to be guarded, making it viral. The ability to parse the interface successfully is the most important aspect of this. Preventing issues with the code generated by the older compiler is definitely important but secondary as usually a client building with the older compiler are not using any of the newer declarations from the newer interface. It's a temporary configuration that we're just trying to make sure is possible to use for clients with existing, shipping code.

nickolas-pohilets replied:
> But that gets tricky because other references to the class elsewhere in the interface would also need to be guarded, making it viral.

Not sure if I understood this bit.

I meant something like this:
```swift
#if $IsolatedDeinit
open class SomeClass {
  isolated deinit
}
#else
public class SomeClass {
}
#endif

func foo(_ x: SomeClass)
```

For example, why would reference in `foo()` need to be guarded?

nickolas-pohilets replied:
@tshortli, implemented, please take a look

tshortli replied:
> Not sure if I understood this bit.

Sorry, what I said didn't make sense because I read too quickly and thought you were going to hide the class definition of the class entirely when the feature is unavailable, in which case you have to consider what happens to references to the class in other declarations. The approach of preventing subclassing when the feature is unavailable is a clever one that should work nicely here.

I'll take a look at the updates to the PR in more depth tomorrow.

ahoppen commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -8571,6 +8571,11 @@ parseDeclDeinit(ParseDeclOptions Flags, DeclAttributes &Attributes) {
   };
+    
+  SourceLoc asyncLoc, throwsLoc;
+  parseEffectsSpecifiers(SourceLoc(),
+                         asyncLoc, /*reasync=*/nullptr,
+                         throwsLoc, /*rethrows=*/nullptr);
```
You will also need to update the new parser to accept an `async` specifier after `deist`.

nickolas-pohilets replied:
Draft - https://github.com/apple/swift-syntax/pull/1656

nickolas-pohilets committed:
Enabled isolated deinit in Sema

nickolas-pohilets committed:
Introduced isolated deinit as a SIL entity

nickolas-pohilets committed:
SILGen for isolated deinit

nickolas-pohilets committed:
Generating errors when isolation of the subclass deinit is not compatible with superclass deinit

nickolas-pohilets committed:
Added more tests. Implemented logic for inferring actor isolation for deinit from base class and skipping isolation when deinit is implicit.

nickolas-pohilets committed:
Added checks for mangled names for isolated deinits

nickolas-pohilets committed:
Updated SILGen for isolated deinit in @objc classes

Failing: Concurrency/flow_isolation.swift

nickolas-pohilets committed:
Fixed a bug in computing isolation of @objc actors. Added tests for @objc declarations

nickolas-pohilets committed:
Import dealloc method as deinit

nickolas-pohilets committed:
Don't generate isolating destructor if dealloc was explicitly isolated in ObjC

nickolas-pohilets committed:
Removed declaration of the AdHocJob from the ABI headers

nickolas-pohilets committed:
Fixed some tests. Using extract_executor SIL instruction instead of custom code.

nickolas-pohilets committed:
Fixed linkage logic for isolated deallocating deinit

nickolas-pohilets committed:
Added isolated deallocator to TBDGen

nickolas-pohilets committed:
Don't hop to @MainActor if already running on the main thread.

nickolas-pohilets committed:
Reviewed code comments. Fixed typos and addressed some TODOs

nickolas-pohilets committed:
Removed some dead code

nickolas-pohilets committed:
Updated Distributed/Runtime/distributed_actor_deinit.swift to check for isolation

nickolas-pohilets committed:
Fixed some typos

nickolas-pohilets committed:
Removed explicit deinit from MainActor - not needed.

nickolas-pohilets committed:
Don't isolate deinit of remote actor proxies

nickolas-pohilets committed:
Expanded test for SIL for distributed actor deinit, added isolated cases

nickolas-pohilets committed:
Don't add isolation attributes for deinit when they were inferred from overriden declaration.

But if isolation was inferred from class isolation and matches with isolation of the overriden declaration, attributes are added.

nickolas-pohilets committed:
Don't switch threads when running isolated deinit of the default actor.

nickolas-pohilets committed:
Rename performOnExecutor into deinitOnExecutor.

It cannot be used for executing general-purpose work, because such function would need to have a different signature to pass isolated actor instance.

And being explicit about using this method only for deinit allows to use object pointer for comparison with executor identity.

nickolas-pohilets committed:
Updated unit tests

nickolas-pohilets committed:
Added more tests for isolated actor deinit:

* Recursive deallocation with multiple locks held
* (UB) Escaping self triggers assertion in stdlib

nickolas-pohilets committed:
Fixed tests on Linux, Windows and older MacOS versions

nickolas-pohilets committed:
Make deinit non-isolated by default

Failing: Distributed/Runtime/distributed_actor_deinit.swift

nickolas-pohilets committed:
Allow accessing let properties in deinit isolated on different actor

nickolas-pohilets committed:
Added flags to be able to control task-locals behaviour in the future

nickolas-pohilets committed:
Implemented suppressing isolated deinit for compatibility with older compiler versions

nickolas-pohilets committed:
Keep current task executor when taking fast path in isolated deinit

nickolas-pohilets committed:
Disable string processing

nickolas-pohilets committed:
Don't read current task executor to avoid extra access of the task-local storage.

nickolas-pohilets committed:
Added unit-tests for (not)copying task-local values. Fixed release after deallocation.

nickolas-pohilets committed:
Added emitting precondition check in the isolated deinit

nickolas-pohilets committed:
Make sure legacy executor checking mode is always used in swift_task_deinitOnExecutor()

nickolas-pohilets committed:
Merge remote-tracking branch 'upstream/main' into mpokhylets/isolated-deinit

# Conflicts:
#	include/swift/Basic/Features.def
#	lib/SILGen/SILGenDestructor.cpp
#	test/Concurrency/flow_isolation.swift
#	test/abi/macOS/arm64/concurrency.swift
#	test/abi/macOS/x86_64/concurrency.swift

nickolas-pohilets commented:
Rebased and updated implementation:
* Removed all code changes related to TL, but updated and preserved unit-tests.
* Preserved `flags` parameter, as an extension point for the future. No strong opinion about this, I don't mind removing it.
* Removed changes for async deinit to separate branch (can be recovered later).
* Removed changes to python tests (out of scope, see also https://github.com/swiftlang/swift/pull/61510).
* Allowed `@preconcurrency` attribute on deist
* Fixed a bug in slow path of the isolated deinit, by removing pair of `swift_retain()/`swift_release()` around `defaultActorDrain()`.
* Always use `Legacy_NoCheckIsolated_NonCrashing` when checking for current executor, even in Swift 6.

Branch is ready to run CI and resume code review process.

hborla commented:
@swift-ci please build toolchain

hborla commented:
@swift-ci please test

nickolas-pohilets committed:
Fixed CI builds

hborla commented:
@swift-ci please build toolchain

hborla commented:
@swift-ci please test

nickolas-pohilets committed:
Temporary disabled test failing on Linux

hborla commented:
@swift-ci please build toolchain

nickolas-pohilets committed:
Reject isolation attributes on deinit in value types

nickolas-pohilets committed:
Using REQUIRES: swift_stdlib_asserts as a cleaner way to disable actor_deinit_escaping_self on Linux

nickolas-pohilets committed:
Updated implementation to handle SWIFT_CONCURRENCY_ACTORS_AS_LOCKS

nickolas-pohilets committed:
Deterministically crash in DefaultActorImpl::destroy() when self escapes in deinit

nickolas-pohilets committed:
Merge branch 'main' into mpokhylets/isolated-deinit

# Conflicts:
#	lib/SILGen/SILGenDistributed.cpp
#	lib/Sema/TypeCheckConcurrency.cpp

ktoso commented:
@swift-ci please smoke test

nickolas-pohilets commented on stdlib/public/Concurrency/Actor.cpp:
```diff
@@ -1642,15 +1647,28 @@ void ProcessOutOfLineJob::process(Job *job) {
 
 void DefaultActorImpl::destroy() {
+  HeapObject *object = asAbstract(this);
+  size_t retainCount = swift_retainCount(object);
```
@kubamracek, I need a bit of your help here.

This change causes linker error when running in `test/embedded/concurrency-actors.swift`:

```text
ld: Undefined symbols:
  _swift_retainCount, referenced from:
      _swift_defaultActor_destroy in libswift_Concurrency.a[4](Actor.cpp.o)
```

What is the correct way to fix this test? Which library contains _swift_retainCount for embedded configuration?

nickolas-pohilets replied:
Looks like `swift_retainCount()` must be handled in `SILLinker::linkEmbeddedRuntimeFromStdlib()`. This can be achieved by handling `swift_retainCount()` ad-hoc (not affecting other 3 use cases of the `RuntimeFunctions.def`) or by adding `swift_retainCount()` to `RuntimeFunctions.def` (for consistency). @kubamracek, @eeckstein, which one do you think is a better approach?

eeckstein replied:
I think your change is fine

nickolas-pohilets committed:
Added swift_retainCount to EmbeddedRuntime.swift

Fixes test/embedded/concurrency-actors.swift

nickolas-pohilets committed:
Don't add swift_retainCount() to RuntimeFunctions.def, but explicitly link it in linkEmbeddedRuntimeFromStdlib()

nickolas-pohilets commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -5931,6 +5934,11 @@ ERROR(async_named_decl_must_be_available_from_async,none,
+      
+      
+ERROR(isolated_deinit_no_isolation,none,
+      "deinit is marked isolated, but containing class %0 is not isolated to an actor",
```
Updated `lib/Sema/TypeCheckAttr.cpp` to reject isolation attributes on `deinit` in structs and enums.

nickolas-pohilets commented on lib/SILGen/SILGenDestructor.cpp:
```diff
@@ -29,46 +29,38 @@
+  RegularLocation loc(dd);
+  loc.markAutoGenerated();
 
   auto cd = cast<ClassDecl>(dd->getDeclContext()->getSelfNominalTypeDecl());
```
`~Copyable` structs are handled in `emitDeallocatingMoveOnlyDestructor()` which does not call this function. Should be ok.

nickolas-pohilets commented on lib/Demangling/Demangler.cpp:
```diff
@@ -3960,6 +3960,10 @@ NodePointer Demangler::demangleFunctionEntity() {
+    case 'Z':
+      Args = None;
+      Kind = Node::Kind::IsolatedDeallocator;
+      break;
```
The same style is used in `P` and `W` cases, so IMO changing this does not help much. But, if you insist, I can make it into a single line. Or did you mean a different style?

ktoso replied:
Yeah not too important

nickolas-pohilets commented on stdlib/public/Concurrency/Actor.cpp:
```diff
@@ -1642,9 +1647,6 @@ void ProcessOutOfLineJob::process(Job *job) {
   delete self;
-
-  // Balances with the swift_release in defaultActorDrain()
-  swift_retain(actor);
```
It is ok for actor to be deallocated while job is running on its executor, logic around `Zombie_ReadyForDeallocation` state takes care of this. But looks like this scenario was never covered by the unit tests. With isolated deinit is became possible to reproduce this scenario reliably.

One of the unit tests that I've added covers this, and detected a bug - actor was deallocated with +1 retain count triggering `fatalError()`. If control could reach `swift_release()`, it would be called on deallocated pointer.

ktoso replied:
Sounds good, thanks for the explanation

nickolas-pohilets commented on test/Concurrency/Runtime/actor_recursive_deinit.swift:
```diff
@@ -0,0 +1,75 @@
+}
+
+func isCurrent(_ a: AnyActor) -> Bool {
+  return isCurrentExecutor(getExecutor(a))
```
I've updated couple tests to use the env variable, and for remaining ones crashing on `false` is ok - it also works as a way of failing the test.

ktoso replied:
Sounds good 👍

nickolas-pohilets committed:
Revert allowing @preconcurrency on deinit declarations

Magically fixed

nickolas-pohilets committed:
Disabled isolated deinit for SWIFT_CONCURRENCY_ACTORS_AS_LOCKS in the runtime

nickolas-pohilets committed:
Implemented zombie actor state in SWIFT_CONCURRENCY_ACTORS_AS_LOCKS mode using additional reference counter

nickolas-pohilets commented on include/swift/AST/DeclAttr.def:
```diff
@@ -326,7 +326,7 @@ SIMPLE_DECL_ATTR(_noAllocation, NoAllocation,
   124)
 SIMPLE_DECL_ATTR(preconcurrency, Preconcurrency,
-  OnFunc | OnConstructor | OnProtocol | OnGenericType | OnVar | OnSubscript | OnEnumElement | OnImport | ABIStableToAdd | ABIBreakingToRemove | APIBreakingToAdd | APIBreakingToRemove,
+  OnFunc | OnConstructor | OnDestructor | OnProtocol | OnGenericType | OnVar | OnSubscript | OnEnumElement | OnImport | ABIStableToAdd | ABIBreakingToRemove | APIBreakingToAdd | APIBreakingToRemove,
```
Seems that `@preconcurrency` does not get added to the `deinit` anymore, so this change is not needed. Reverted.

nickolas-pohilets committed:
Merge branch 'main' into mpokhylets/isolated-deinit

# Conflicts:
#	include/swift/Basic/Features.def
#	lib/AST/ASTPrinter.cpp
#	lib/AST/FeatureSet.cpp

ktoso commented:
@swift-ci please smoke test

ktoso committed:
Merge branch 'main' into mpokhylets/isolated-deinit

ktoso commented:
@swift-ci please smoke test

nickolas-pohilets committed:
Fixed test/SourceKit/InterfaceGen/gen_swift_module.swift

nickolas-pohilets committed:
Merge branch 'main' into mpokhylets/isolated-deinit

ktoso commented:
@swift-ci please smoke test

nickolas-pohilets committed:
Fixed test/Distributed/Runtime/distributed_actor_deinit.swift in optimized mode
